Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -488,11 +375,118 @@ std::pair<s32, s32> Mixer::MixerFifo::GetVolume() const
+
+  m_queue[head] = granule;
+  m_queue_head.store(next_head, std::memory_order_release);
+
+  m_queue_looping.store(false, std::memory_order_relaxed);
+}
+
+void Mixer::MixerFifo::dequeue(Granule& granule)
+{
+  const size_t tail = m_queue_tail.load(std::memory_order_relaxed);
+  size_t next_tail = (tail + 1) % GRANULE_QUEUE_SIZE;
+
+  if (next_tail == m_queue_head.load(std::memory_order_acquire))
+  {
+    if (m_mixer->m_audio_fill_gaps)
+    {
+      next_tail = (tail + GRANULE_QUEUE_SIZE / 2) % GRANULE_QUEUE_SIZE;
+      m_queue_looping.store(true, std::memory_order_relaxed);
+    }
+    else
+    {
+      granule = Granule();
+      return;
+    }
+  }
+
+  if (m_queue_looping.load(std::memory_order_relaxed))
+    m_queue_fade_index = std::min(m_queue_fade_index + 1, FADE_WINDOW.size() - 1);
+  else
+    m_queue_fade_index = 0;
+
+  granule = m_queue[tail];
+  granule.mul(FADE_WINDOW[m_queue_fade_index]);
+
+  m_queue_tail.store(next_tail, std::memory_order_release);
+}
+
+// StereoPair methods (part of MixerFifo)
+
+inline Mixer::MixerFifo::StereoPair
+Mixer::MixerFifo::StereoPair::operator+(const StereoPair& other) const
+{
+  return StereoPair(std::clamp(int(l) + other.l, -0x7fff, 0x7fff),
+                    std::clamp(int(r) + other.r, -0x7fff, 0x7fff));
+}
+
+inline Mixer::MixerFifo::StereoPair
+Mixer::MixerFifo::StereoPair::interp(const StereoPair& s0, const StereoPair& s1,
+                                     const StereoPair& s2, const StereoPair& s3,
+                                     const StereoPair& s4, const StereoPair& s5, int t)
+{
+  const int t0 = 1 << 16;
+  const int t1 = t;
+  const int t2 = (t1 * t1) >> 16;
+  const int t3 = (t1 * t2) >> 16;
+
+  const int c0 = (+0 * t0 + 1 * t1 - 2 * t2 + 1 * t3) / 24;
+  const int c1 = (+0 * t0 - 8 * t1 + 15 * t2 - 7 * t3) / 24;
+  const int c2 = (+3 * t0 + 0 * t1 - 7 * t2 + 4 * t3) / 6;
+  const int c3 = (+0 * t0 + 2 * t1 + 5 * t2 - 4 * t3) / 6;
+  const int c4 = (+0 * t0 - 1 * t1 - 6 * t2 + 7 * t3) / 24;
+  const int c5 = (+0 * t0 + 0 * t1 + 1 * t2 - 1 * t3) / 24;
+
+  const int l = (c0 * s0.l + c1 * s1.l + c2 * s2.l + c3 * s3.l + c4 * s4.l + c5 * s5.l) >> 15;
+  const int r = (c0 * s0.r + c1 * s1.r + c2 * s2.r + c3 * s3.r + c4 * s4.r + c5 * s5.r) >> 15;
+
+  return StereoPair(short(std::clamp<int>(l, -0x7fff, 0x7fff)),
+                    short(std::clamp<int>(r, -0x7fff, 0x7fff)));
+}
+
+// Implementation of Granule's constructor
+Mixer::MixerFifo::Granule::Granule(const GranuleBuffer& input, size_t start_index)
+{
+  auto buffer_middle = m_buffer.begin() + start_index;
+  auto input_middle = input.end() - start_index;
+
+  std::copy(input_middle, input.end(), m_buffer.begin());
+  std::copy(input.begin(), input_middle, buffer_middle);
+
+  for (size_t i = 0; i < m_buffer.size(); ++i)
+    m_buffer[i].mul(GRANULE_WINDOW[i]);
+}
+
+// Implementation of the operator[]
+Mixer::MixerFifo::StereoPair Mixer::MixerFifo::Granule::operator[](uint32_t frac) const
```

Comment:
I think the calling code would be easier to follow if this wasn't an `operator[]`. It's not immediately obvious that `[]` would be doing all these calculations, generally `[]` would just be accessing some container. Maybe a member-function named `InterpStereoPair` or whatever you think is appropriate, I'm not a sound guy.